### DISCUSSION TITLE
[FE102] - JavaScript 與瀏覽器的溝通 - 改變元素內容的 inner&outer ，關於 innerText & innerHTML 和 outerHTML 印出來的東西會有差異的問題

### DISCUSSION THREAD

**Discussion by KW0409** [Lidemy/forum#107] [前端相關問題]
大家好今天在自己的編輯器上嘗試了這個影片中學到的三種改變元素內容的方法
分別是 innerText,innerHTML&outerHTML
發現在使用 innerText 和 innerHTML 改變內容之後
再去看網頁的畫面和印出的元素，
都會是改變後的內容沒錯
如下圖
<h3>這是用 innerText</h3>
<img src='https://user-images.githubusercontent.com/80152099/118358090-da38ec80-b5af-11eb-8fd8-a185a4a15c13.png' title='這是用 innerText' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358180-3ef44700-b5b0-11eb-99fc-5cef765815c3.png' title='這是用 innerText' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358188-4a477280-b5b0-11eb-8a2a-334decc6e430.png'
 title='這是用 innerText' width='80%' height='80%' />

<h3>這是用 innerHTML</h3>
<img src='https://user-images.githubusercontent.com/80152099/118358484-90510600-b5b1-11eb-8f1c-c35dd20ec24e.png'  title='這是用 innerHTML' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358503-b080c500-b5b1-11eb-9d34-bd619a05fbb8.png'  title='這是用 innerHTML' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358507-b2e31f00-b5b1-11eb-87f1-04a321c9598c.png'  title='這是用 innerHTML' width='80%' height='80%' />

但是用 outerHTML 改變元素的內容之後
再去看網頁的內容和印出的元素，
發現網頁的畫面會是改變後的內容沒錯
但印出的元素
卻是還沒改變前的元素內容
如下圖
<h3>這是用 outerHTML</h3>
<img src='https://user-images.githubusercontent.com/80152099/118358646-6cda8b00-b5b2-11eb-9dbe-3470c748bac4.png'  title='這是用 outerHTML' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358648-6ea44e80-b5b2-11eb-9178-4a0778863885.png' title='這是用 outerHTML' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358651-6fd57b80-b5b2-11eb-9a80-adc919caa846.png'  title='這是用 outerHTML' width='80%' height='80%' />

本來想説可能是因為原本選到的&lt;p&gt;被改掉，程式碼可能因為找不到符合 .box > p 的條件
所以才會印出原本的元素及內容
所以改成用同樣的標籤&lt;p&gt;去改變，但結果還是一樣
<img src='https://user-images.githubusercontent.com/80152099/118358810-28032400-b5b3-11eb-9996-87bae02ff6ef.png'  title='這是用 outerHTML' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358816-29cce780-b5b3-11eb-92c1-b599ce3a7363.png'  title='這是用 outerHTML' width='80%' height='80%' />
<img src='https://user-images.githubusercontent.com/80152099/118358818-2b96ab00-b5b3-11eb-8cf9-ef263189c96c.png'  title='這是用 outerHTML' width='80%' height='80%' />

想知道為什麼會有這樣的差異，懇請大家幫忙解惑～

**Comment by cwc329** [comment]
你可以看看這個：
https://pjchender.blogspot.com/2016/03/javascriptby-referenceby-value.html
https://blog.techbridge.cc/2018/06/23/javascript-call-by-value-or-reference/

**Reply by KW0409** [reply]
感謝提供的文章
我看完之後的理解是 outerHTML 的方式比較像 call by value，
而 innerText 和 innerHTML 比較像 call by reference(應該沒有理解錯....吧，如果有錯的話，歡迎指正)
但是為什麼會有這樣的區別呢？
是因為 innerText 和 innerHTML 是針對其中的一部份作改變，就像是 obj.2 = x
而 outerHTML 因為是針對整體去做改變，就像是 obj = {2: x}
所以一個會改變到元素本身，另一個其實沒有改變到元素本身是這樣嗎？
